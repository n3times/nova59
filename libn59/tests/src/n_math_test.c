#include "n_test.h"

#include <math.h>
#include <stdio.h>

static void p(n_t n, n_t res, n_err_t err, char *op) {
  char s[N_PRINT_MAX_SIZE];
  char s_res[N_PRINT_MAX_SIZE];

  printf("%s %s    =   %s%s\n",
         n_print(n, s),
         op,
         n_print(res, s_res),
         err ? "?" : "");
}

static void test_funs() {
  typedef n_t (*fun_t)(n_t n, n_err_t *err);

  n_err_t err;
  n_t res;
  n_t ns[] = { TEST_NUMBERS_2 };
  fun_t funs[] = { n_square, n_sqrt, n_log, n_ln, n_1_x, n_exp, n_pow10 };
  char *fun_strs[] =
      { "  ^2", "  vx", " log", "  ln", " 1/x" , " exp", " ^10" };

  // Test all the functions for all the test numbers.
  for (int i = 0; i < N_ELEMS(funs); i++) {
    for (int j = 0; j < N_ELEMS(ns); j++) {
      res = funs[i](ns[j], &err);
      p(ns[j], res, err, fun_strs[i]);
    }
    printf("\n\n");
  }
}

static void test_funs_2() {
  typedef n_t (*fun_t)(n_t n);

  n_t res;
  n_t ns[] = { TEST_NUMBERS_2 };
  fun_t funs[] = { n_int, n_frac, n_abs, n_chs };
  char *ops[] = { " int", " frc", " abs", " chs" };

  // Test all the functions for all the test numbers.
  for (int i = 0; i < N_ELEMS(funs); i++) {
    for (int j = 0; j < N_ELEMS(ns); j++) {
      res = funs[i](ns[j]);
      p(ns[j], res, N_ERR_NONE, ops[i]);
    }
    printf("\n\n");
  }
}

static void test_int() {
  n_t res;

  for (int i = -1; i < 14; i++) {
    res = n_make(pow(10, i));
    p(res, n_int(res), N_ERR_NONE, " int");
  }
  printf("\n");

  for (int i = -2; i < 14; i++) {
    res = n_make(9.999999999999 * pow(10, i));
    p(res, n_int(res), N_ERR_NONE, " int");
  }
  printf("\n");

  for (int i = -1; i < 14; i++) {
    res = n_make(1.234567890123 * pow(10, i));
    p(res, n_int(res), N_ERR_NONE, " int");
  }
  printf("\n");

  for (int i = -1; i < 14; i++) {
    res = n_make(-1.234567890123 * pow(10, i));
    p(res, n_int(res), N_ERR_NONE, " int");
  }
  printf("\n");
}

static void test_frac() {
  n_t res;

  for (int i = -1; i < 14; i++) {
    res = n_make(pow(10, i));
    p(res, n_frac(res), N_ERR_NONE, " frc");
  }
  printf("\n");

  for (int i = -2; i < 14; i++) {
    res = n_make(9.999999999999 * pow(10, i));
    p(res, n_frac(res), N_ERR_NONE, " frc");
  }
  printf("\n");

  for (int i = -1; i < 14; i++) {
    res = n_make(1.234567890123 * pow(10, i));
    p(res, n_frac(res), N_ERR_NONE, " frc");
  }
  printf("\n");

  for (int i = -1; i < 14; i++) {
    res = n_make(-1.234567890123 * pow(10, i));
    p(res, n_frac(res), N_ERR_NONE, " frc");
  }
}

static void test_sqrt() {
  n_t n_1_plus_eps  = n_make(1.000000000001);
  n_t n_1_minus_eps = n_make(.9999999999999);

  p(n_1_plus_eps,  n_sqrt(n_1_plus_eps, NULL),  N_ERR_NONE, "sqrt");
  p(n_1_minus_eps, n_sqrt(n_1_minus_eps, NULL), N_ERR_NONE, "sqrt");
}

int main() {
  test_funs();

  test_funs_2();

  test_int();

  test_frac();

  printf("\n");
  test_sqrt();
}
