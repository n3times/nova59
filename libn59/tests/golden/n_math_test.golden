 0000000000000 00   ^2    =    0000000000000 00
 1000000000000 00   ^2    =    1000000000000 00
 3141592653590 00   ^2    =    9869604401090 00
 1000000000000-99   ^2    =    1000000000000-99?
 9999999999999 99   ^2    =    9999999999999 99?
-1000000000000 00   ^2    =    1000000000000 00
-3141592653590 00   ^2    =    9869604401090 00
-1000000000000-99   ^2    =    1000000000000-99?
-9999999999999 99   ^2    =    9999999999999 99?
 1000000000000 01   ^2    =    1000000000000 02
 2718281828459 00   ^2    =    7389056098930 00
-1000000000000 01   ^2    =    1000000000000 02
-2718281828459 00   ^2    =    7389056098930 00


 0000000000000 00   vx    =    0000000000000 00
 1000000000000 00   vx    =    1000000000000 00
 3141592653590 00   vx    =    1772453850906 00
 1000000000000-99   vx    =    3162277660168-50
 9999999999999 99   vx    =    9999999999999 49
-1000000000000 00   vx    =    1000000000000 00?
-3141592653590 00   vx    =    1772453850906 00?
-1000000000000-99   vx    =    3162277660168-50?
-9999999999999 99   vx    =    9999999999999 49?
 1000000000000 01   vx    =    3162277660168 00
 2718281828459 00   vx    =    1648721270700 00
-1000000000000 01   vx    =    3162277660168 00?
-2718281828459 00   vx    =    1648721270700 00?


 0000000000000 00  log    =   -9999999999999 99?
 1000000000000 00  log    =    0000000000000 00
 3141592653590 00  log    =    4971498726942-01
 1000000000000-99  log    =   -9900000000000 01
 9999999999999 99  log    =    1000000000000 02
-1000000000000 00  log    =    0000000000000 00?
-3141592653590 00  log    =    4971498726942-01?
-1000000000000-99  log    =   -9900000000000 01?
-9999999999999 99  log    =    1000000000000 02?
 1000000000000 01  log    =    1000000000000 00
 2718281828459 00  log    =    4342944819032-01
-1000000000000 01  log    =    1000000000000 00?
-2718281828459 00  log    =    4342944819032-01?


 0000000000000 00   ln    =   -9999999999999 99?
 1000000000000 00   ln    =    0000000000000 00
 3141592653590 00   ln    =    1144729885849 00
 1000000000000-99   ln    =   -2279559242064 02
 9999999999999 99   ln    =    2302585092994 02
-1000000000000 00   ln    =    0000000000000 00?
-3141592653590 00   ln    =    1144729885849 00?
-1000000000000-99   ln    =   -2279559242064 02?
-9999999999999 99   ln    =    2302585092994 02?
 1000000000000 01   ln    =    2302585092994 00
 2718281828459 00   ln    =    1000000000000 00
-1000000000000 01   ln    =    2302585092994 00?
-2718281828459 00   ln    =    1000000000000 00?


 0000000000000 00  1/x    =    9999999999999 99?
 1000000000000 00  1/x    =    1000000000000 00
 3141592653590 00  1/x    =    3183098861838-01
 1000000000000-99  1/x    =    1000000000000 99
 9999999999999 99  1/x    =    1000000000000-99?
-1000000000000 00  1/x    =   -1000000000000 00
-3141592653590 00  1/x    =   -3183098861838-01
-1000000000000-99  1/x    =   -1000000000000 99
-9999999999999 99  1/x    =   -1000000000000-99?
 1000000000000 01  1/x    =    1000000000000-01
 2718281828459 00  1/x    =    3678794411714-01
-1000000000000 01  1/x    =   -1000000000000-01
-2718281828459 00  1/x    =   -3678794411714-01


 0000000000000 00  exp    =    1000000000000 00
 1000000000000 00  exp    =    2718281828459 00
 3141592653590 00  exp    =    2314069263278 01
 1000000000000-99  exp    =    1000000000000 00
 9999999999999 99  exp    =    9999999999999 99?
-1000000000000 00  exp    =    3678794411714-01
-3141592653590 00  exp    =    4321391826377-02
-1000000000000-99  exp    =    1000000000000 00
-9999999999999 99  exp    =    1000000000000-99?
 1000000000000 01  exp    =    2202646579480 04
 2718281828459 00  exp    =    1515426224148 01
-1000000000000 01  exp    =    4539992976249-05
-2718281828459 00  exp    =    6598803584532-02


 0000000000000 00  ^10    =    1000000000000 00
 1000000000000 00  ^10    =    1000000000000 01
 3141592653590 00  ^10    =    1385455731368 03
 1000000000000-99  ^10    =    1000000000000 00
 9999999999999 99  ^10    =    9999999999999 99?
-1000000000000 00  ^10    =    1000000000000-01
-3141592653590 00  ^10    =    7217841590744-04
-1000000000000-99  ^10    =    1000000000000 00
-9999999999999 99  ^10    =    1000000000000-99?
 1000000000000 01  ^10    =    1000000000000 10
 2718281828459 00  ^10    =    5227352996704 02
-1000000000000 01  ^10    =    1000000000000-10
-2718281828459 00  ^10    =    1913014102225-03


 0000000000000 00  int    =    0000000000000 00
 1000000000000 00  int    =    1000000000000 00
 3141592653590 00  int    =    3000000000000 00
 1000000000000-99  int    =    0000000000000 00
 9999999999999 99  int    =    9999999999999 99
-1000000000000 00  int    =   -1000000000000 00
-3141592653590 00  int    =   -3000000000000 00
-1000000000000-99  int    =    0000000000000 00
-9999999999999 99  int    =   -9999999999999 99
 1000000000000 01  int    =    1000000000000 01
 2718281828459 00  int    =    2000000000000 00
-1000000000000 01  int    =   -1000000000000 01
-2718281828459 00  int    =   -2000000000000 00


 0000000000000 00  frc    =    0000000000000 00
 1000000000000 00  frc    =    0000000000000 00
 3141592653590 00  frc    =    1415926535900-01
 1000000000000-99  frc    =    1000000000000-99
 9999999999999 99  frc    =    0000000000000 00
-1000000000000 00  frc    =    0000000000000 00
-3141592653590 00  frc    =   -1415926535900-01
-1000000000000-99  frc    =   -1000000000000-99
-9999999999999 99  frc    =    0000000000000 00
 1000000000000 01  frc    =    0000000000000 00
 2718281828459 00  frc    =    7182818284590-01
-1000000000000 01  frc    =    0000000000000 00
-2718281828459 00  frc    =   -7182818284590-01


 0000000000000 00  abs    =    0000000000000 00
 1000000000000 00  abs    =    1000000000000 00
 3141592653590 00  abs    =    3141592653590 00
 1000000000000-99  abs    =    1000000000000-99
 9999999999999 99  abs    =    9999999999999 99
-1000000000000 00  abs    =    1000000000000 00
-3141592653590 00  abs    =    3141592653590 00
-1000000000000-99  abs    =    1000000000000-99
-9999999999999 99  abs    =    9999999999999 99
 1000000000000 01  abs    =    1000000000000 01
 2718281828459 00  abs    =    2718281828459 00
-1000000000000 01  abs    =    1000000000000 01
-2718281828459 00  abs    =    2718281828459 00


 0000000000000 00  chs    =    0000000000000 00
 1000000000000 00  chs    =   -1000000000000 00
 3141592653590 00  chs    =   -3141592653590 00
 1000000000000-99  chs    =   -1000000000000-99
 9999999999999 99  chs    =   -9999999999999 99
-1000000000000 00  chs    =    1000000000000 00
-3141592653590 00  chs    =    3141592653590 00
-1000000000000-99  chs    =    1000000000000-99
-9999999999999 99  chs    =    9999999999999 99
 1000000000000 01  chs    =   -1000000000000 01
 2718281828459 00  chs    =   -2718281828459 00
-1000000000000 01  chs    =    1000000000000 01
-2718281828459 00  chs    =    2718281828459 00


 1000000000000-01  int    =    0000000000000 00
 1000000000000 00  int    =    1000000000000 00
 1000000000000 01  int    =    1000000000000 01
 1000000000000 02  int    =    1000000000000 02
 1000000000000 03  int    =    1000000000000 03
 1000000000000 04  int    =    1000000000000 04
 1000000000000 05  int    =    1000000000000 05
 1000000000000 06  int    =    1000000000000 06
 1000000000000 07  int    =    1000000000000 07
 1000000000000 08  int    =    1000000000000 08
 1000000000000 09  int    =    1000000000000 09
 1000000000000 10  int    =    1000000000000 10
 1000000000000 11  int    =    1000000000000 11
 1000000000000 12  int    =    1000000000000 12
 1000000000000 13  int    =    1000000000000 13

 9999999999999-02  int    =    0000000000000 00
 9999999999999-01  int    =    0000000000000 00
 9999999999999 00  int    =    9000000000000 00
 9999999999999 01  int    =    9900000000000 01
 9999999999999 02  int    =    9990000000000 02
 9999999999999 03  int    =    9999000000000 03
 9999999999999 04  int    =    9999900000000 04
 9999999999999 05  int    =    9999990000000 05
 9999999999999 06  int    =    9999999000000 06
 9999999999999 07  int    =    9999999900000 07
 9999999999999 08  int    =    9999999990000 08
 9999999999999 09  int    =    9999999999000 09
 9999999999999 10  int    =    9999999999900 10
 9999999999999 11  int    =    9999999999990 11
 9999999999999 12  int    =    9999999999999 12
 9999999999999 13  int    =    9999999999999 13

 1234567890123-01  int    =    0000000000000 00
 1234567890123 00  int    =    1000000000000 00
 1234567890123 01  int    =    1200000000000 01
 1234567890123 02  int    =    1230000000000 02
 1234567890123 03  int    =    1234000000000 03
 1234567890123 04  int    =    1234500000000 04
 1234567890123 05  int    =    1234560000000 05
 1234567890123 06  int    =    1234567000000 06
 1234567890123 07  int    =    1234567800000 07
 1234567890123 08  int    =    1234567890000 08
 1234567890123 09  int    =    1234567890000 09
 1234567890123 10  int    =    1234567890100 10
 1234567890123 11  int    =    1234567890120 11
 1234567890123 12  int    =    1234567890123 12
 1234567890123 13  int    =    1234567890123 13

-1234567890123-01  int    =    0000000000000 00
-1234567890123 00  int    =   -1000000000000 00
-1234567890123 01  int    =   -1200000000000 01
-1234567890123 02  int    =   -1230000000000 02
-1234567890123 03  int    =   -1234000000000 03
-1234567890123 04  int    =   -1234500000000 04
-1234567890123 05  int    =   -1234560000000 05
-1234567890123 06  int    =   -1234567000000 06
-1234567890123 07  int    =   -1234567800000 07
-1234567890123 08  int    =   -1234567890000 08
-1234567890123 09  int    =   -1234567890000 09
-1234567890123 10  int    =   -1234567890100 10
-1234567890123 11  int    =   -1234567890120 11
-1234567890123 12  int    =   -1234567890123 12
-1234567890123 13  int    =   -1234567890123 13

 1000000000000-01  frc    =    1000000000000-01
 1000000000000 00  frc    =    0000000000000 00
 1000000000000 01  frc    =    0000000000000 00
 1000000000000 02  frc    =    0000000000000 00
 1000000000000 03  frc    =    0000000000000 00
 1000000000000 04  frc    =    0000000000000 00
 1000000000000 05  frc    =    0000000000000 00
 1000000000000 06  frc    =    0000000000000 00
 1000000000000 07  frc    =    0000000000000 00
 1000000000000 08  frc    =    0000000000000 00
 1000000000000 09  frc    =    0000000000000 00
 1000000000000 10  frc    =    0000000000000 00
 1000000000000 11  frc    =    0000000000000 00
 1000000000000 12  frc    =    0000000000000 00
 1000000000000 13  frc    =    0000000000000 00

 9999999999999-02  frc    =    9999999999999-02
 9999999999999-01  frc    =    9999999999999-01
 9999999999999 00  frc    =    9999999999990-01
 9999999999999 01  frc    =    9999999999900-01
 9999999999999 02  frc    =    9999999999000-01
 9999999999999 03  frc    =    9999999990000-01
 9999999999999 04  frc    =    9999999900000-01
 9999999999999 05  frc    =    9999999000000-01
 9999999999999 06  frc    =    9999990000000-01
 9999999999999 07  frc    =    9999900000000-01
 9999999999999 08  frc    =    9999000000000-01
 9999999999999 09  frc    =    9990000000000-01
 9999999999999 10  frc    =    9900000000000-01
 9999999999999 11  frc    =    9000000000000-01
 9999999999999 12  frc    =    0000000000000 00
 9999999999999 13  frc    =    0000000000000 00

 1234567890123-01  frc    =    1234567890123-01
 1234567890123 00  frc    =    2345678901230-01
 1234567890123 01  frc    =    3456789012300-01
 1234567890123 02  frc    =    4567890123000-01
 1234567890123 03  frc    =    5678901230000-01
 1234567890123 04  frc    =    6789012300000-01
 1234567890123 05  frc    =    7890123000000-01
 1234567890123 06  frc    =    8901230000000-01
 1234567890123 07  frc    =    9012300000000-01
 1234567890123 08  frc    =    1230000000000-02
 1234567890123 09  frc    =    1230000000000-01
 1234567890123 10  frc    =    2300000000000-01
 1234567890123 11  frc    =    3000000000000-01
 1234567890123 12  frc    =    0000000000000 00
 1234567890123 13  frc    =    0000000000000 00

-1234567890123-01  frc    =   -1234567890123-01
-1234567890123 00  frc    =   -2345678901230-01
-1234567890123 01  frc    =   -3456789012300-01
-1234567890123 02  frc    =   -4567890123000-01
-1234567890123 03  frc    =   -5678901230000-01
-1234567890123 04  frc    =   -6789012300000-01
-1234567890123 05  frc    =   -7890123000000-01
-1234567890123 06  frc    =   -8901230000000-01
-1234567890123 07  frc    =   -9012300000000-01
-1234567890123 08  frc    =   -1230000000000-02
-1234567890123 09  frc    =   -1230000000000-01
-1234567890123 10  frc    =   -2300000000000-01
-1234567890123 11  frc    =   -3000000000000-01
-1234567890123 12  frc    =    0000000000000 00
-1234567890123 13  frc    =    0000000000000 00

 1000000000001 00 sqrt    =    1000000000000 00
 9999999999999-01 sqrt    =    9999999999999-01
